### PR TITLE
Level up manager

### PR DESCRIPTION
**Pull request by Eein**

### CODE REVIEW

File: memory/level_up_manager.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
+from enum import Enum
+from typing import Self
+
+from memory.core import mem_handle
+from memory.mappers.player_party_character import PlayerPartyCharacter
+
+
+class LevelUpUpgradeType(Enum):
```
Comment (orkaboy) [inline memory/level_up_manager.py:8]: These map directly onto memory values in the game or?

File: memory/level_up_manager.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+# This is actually called LevelUpSceneController, but its a manager
+# class, so we'll call it that here.
+class LevelUpManager:
+    NULL_POINTER = 0xFFFFFFFF
+    ZERO_NULL_POINTER = 0x0
+    ITEM_OBJECT_OFFSET = 0x8
+    ITEM_INDEX_0_ADDRESS = 0x20
+
+    def __init__(self: Self) -> None:
```
Comment (orkaboy) [inline memory/level_up_manager.py:33]: Suggest adding type hints to these, especially the "publicly" exposed values that are assigned `None` or `[]`. So for example:

```python
self.current_upgrades: list[LevelUpUpgrade] = []
```

File: memory/level_up_manager.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            count = self.memory.read_int(items + 0x18)
+            address = self.ITEM_INDEX_0_ADDRESS
+            for _item in range(count):
+                item_ptr = self.memory.follow_pointer(items, [address, 0x0])
+                upgrade_type = LevelUpUpgradeType(self.memory.read_int(item_ptr + 0x38))
+                active_ptr = self.memory.follow_pointer(item_ptr, [0x50, 0x0])
+                active = not self.memory.read_bool(active_ptr + 0xA0)
+                upgrades.append(LevelUpUpgrade(upgrade_type, active))
+                address += self.ITEM_OBJECT_OFFSET
+        self.current_level_up_upgrades = upgrades
```
Comment (orkaboy) [inline memory/level_up_manager.py:115]: shouldn't this be `self.current_upgrades`? Either way, we should not assign a value that isn't declared in `__init__()`

File: GUI/debug_menu.py [hunk 4]
```diff
@@ -41,6 +43,19 @@ def execute(self: Self, top_level: bool) -> bool:
...
+        LayoutHelper.add_spacer()
+        imgui.text_wrapped("Level Up Info")
+        imgui.text_wrapped(
+            f"Level Up Screen Active: {level_up_manager.level_up_screen_active}"
+        )
+        if level_up_manager.level_up_screen_active:
+            imgui.text_wrapped(
+                f"Current Character: {level_up_manager.current_character.value}"
+            )
+            for option in level_up_manager.current_level_up_upgrades:
```
Comment (orkaboy) [inline GUI/debug_menu.py:55]: Change this one as well if the name change in `level_up_manager.py` is done.

File: memory/level_up_manager.py (hunk 1 above)
Comment (Eein) [reply memory/level_up_manager.py:8]: Yes

File: memory/level_up_manager.py (hunk 3 above)
Comment (Eein) [reply memory/level_up_manager.py:115]: Yes - will fix

**Review by orkaboy — APPROVED** [review]

**Pull request merged**